### PR TITLE
Add days_in_year and decimal_year to dt accessor

### PR DESCRIPTION
**Pull request by aulemahal**
- [ ] Closes #xxxx
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

Changes:
- Added `is_leap_year` to the `CFTimeIndex`.
- Added two dt accessors:
    + `days_in_year` : the number of days in the year of the datetime, based on the previous method for cftime and the already existing one for numpy.
   + `decimal_year` : the date as the year + fraction of the elapsed year. The underlying function is used for converting 360_day calendars and for `interp_calendar`. This can be useful for certain astronomical calculations, among others.
- Rewrote `xr.coding.calendar_ops._days_in_year` to make it more explicit and more performant. It now accepts both DataArrays or scalars, returning the appropriate type. This is used for 360 day calendar conversion. I could not use the dt accessor as there is a need for a calendar override here.

### CODE REVIEW

**Review by spencerkclark — COMMENTED** [review]
Thanks @aulemahal—this seems useful.  I realize it adds a bit of extra work, but what do you think about adding `days_in_year` and `decimal_year` attributes to `cftime.datetime` objects themselves?  It seems cleaner for the calendar math / logic to be defined there rather than xarray—then adding accessors is fairly straightforward.

Admittedly it was not totally planned this way, but this is how the `days_in_month` accessor came to be (https://github.com/pydata/xarray/pull/3935).

**Comment by aulemahal** [comment]
I didn't think of that, but indeed, this seems to make most sense. I guess this also means implementing it in `pandas` ? The current  `decimal_year` uses stuff from cftime, but could be rewritten without in pandas.

Well, I'll do that, but it will take more time :sweat_smile:

**Comment by spencerkclark** [comment]
True, yeah, let me ponder it a bit more before we make issues in a bunch of places.  There may at least be a way we can take better advantage of existing public cftime and pandas code.

File: xarray/core/accessor_dt.py [hunk 1]
```diff
@@ -533,6 +534,26 @@ def calendar(self) -> CFCalendar:
...
+
+    @property
+    def decimal_year(self) -> T_DataArray:
+        """Convert the dates as a fractional year."""
+        obj_type = type(self._obj)
+        result = _datetime_to_decimal_year(self._obj)
+        return obj_type(
+            result, name="decimal_year", coords=self._obj.coords,
+            dims=self._obj.dims, attrs=self._obj.attrs
+        )
```
Comment (spencerkclark) [inline xarray/core/accessor_dt.py:555]:
```suggestion
        result = _decimal_year(self._obj)
        newvar = self._obj.variable.copy(data=result, deep=False)
        return self._obj._replace(newvar, name="decimal_year")
```

**Review by spencerkclark — COMMENTED** [review]
Indeed I think there are ways we can simplify a lot of this.  Here are some initial thoughts—let me know if they make sense.

While it would be nice if cftime and pandas supported these as attributes of datetime-like objects, for some of these calendar operations it seems like you still need to have a function that allows you to query how many days are in an integer year (unattached to a datetime) with a given calendar, so for now we might as well use that function in implementing the `days_in_year` accessor.  

It is also nice not to need to worry about only allowing these accessors for certain versions of cftime and pandas.  We could still think about trying to upstream some of this functionality eventually and switching to that once minimum versions were appropriate, but that could come later.

File: xarray/coding/calendar_ops.py [hunk 2]
```diff
@@ -304,10 +320,11 @@ def _datetime_to_decimal_year(times, dim="time", calendar=None):
     """
     from xarray.core.dataarray import DataArray
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xarray/coding/calendar_ops.py]: Sadly `rollback` doesn't touch the time part of the datetime. Ex:  "2001-01-04 14:55:01" -> `rollback YS` -> "2001-01-01 14:55:01". The best solution I found to make this more generic as requested was to re-init an object.

But the `calendar` override seemed non necessary here, so I was able to remove it.

File: xarray/coding/calendar_ops.py [hunk 3]
```diff
@@ -22,14 +24,35 @@
...
+    # A year is a leap year if either (i) it is divisible by 4 but not by 100 or (ii) it is divisible by 400
+    return ((years % 4 == 0) & (years % 100 != 0)) | (years % 400 == 0)
+
+
+def _leap_julian(years):
+    # A year is a leap year if it is divisible by 4, even if it is also divisible by 100
+    return years % 4 == 0
+
+
+def _days_in_year(years, calendar):
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xarray/coding/calendar_ops.py]: I didn't know about that! I implemented `is_leap_year` on `CFTimeIndex` and used that for calculating `days_in_year` on the accessor.

I realized afterwards that `days_in_year` with a calendar override was also needed for  `interp_calendar` so I wrote a second function here too.

**Comment by aulemahal** [comment]
@spencerkclark It took me a long time to come back to this! I tried to address your comments the best I could.

The top post was edited to mirror how I changed this PR. I will try to find time and open PRs in `pandas` and `cftime`.

File: xarray/coding/calendar_ops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (spencerkclark) [reply xarray/coding/calendar_ops.py]: Ah right, I think you could `floor("D")` the times before passing to `_rollback`:
```python
def _decimal_year(times):
    years = times.dt.year
    floored_times = times.dt.floor("D")
    deltas = times - _rollback(floored_times, "YS")
    days_in_years = times.dt.days_in_year.astype("timedelta64[D]")
    return years + deltas / days_in_years
```
But honestly the approach you took is also pretty clean, so I think it is fine to leave as is.

**Review by spencerkclark — COMMENTED** [review]
Thanks @aulemahal!  I think this is close—just a few minor suggestions.  

Could you also add `days_in_year` and `decimal_year` as fields in the `test_dask_field_access` test within `test_accessor_dt.py`?  This will confirm these are / remain appropriately dask and multi-dimensional array compatible.

**Comment by spencerkclark** [comment]
The test failures are all in environments with NumPy < 2, and the issue looks reminiscent of https://github.com/pydata/xarray/issues/9387#issuecomment-2306007750.  The dask issues with the rollback approach were unrelated and seemed more readily solvable, which I tried to address in https://github.com/Ouranosinc/xarray/pull/18.  @aulemahal let me know if those changes make sense.

**Comment by spencerkclark** [comment]
Actually my approach in https://github.com/Ouranosinc/xarray/pull/18 suffers from the same problem as the `_yearstart_np` approach, since `_rollback` similarly returns a `datetime64[ns]` array 🤦.  I'll see if I can come up with a different workaround, but worst case we can skip the dask test with NumPy < 2.

**Comment by spencerkclark** [comment]
OK I think I have an alternative approach that should work now in https://github.com/Ouranosinc/xarray/pull/18 (basically keeping the decimal year computation fully within NumPy and not exposing `datetime64[ns]` arrays as dask outputs).  

I tested this interactively for cftime dates too, but I'll work on adding some official tests for this.

**Comment by spencerkclark** [comment]
I updated the tests for `decimal_year` in https://github.com/Ouranosinc/xarray/pull/18 to now cover all combinations of `datetime64[ns]` / cftime and NumPy / dask.

**Review by spencerkclark — APPROVED** [review]
Thanks for your patience @aulemahal—this looks good to me now!

**Pull request merged**